node/src/components/contract_runtime/rewards.rs

```diff
@@ -169,203 +278,151 @@ impl ErasInfo {
-        });
-
-        for block in blocks_from_current_era {
+        // Collect all rewards as a ratio:
+        for block in rewards_info.blocks_from_era(current_era_id) {
             // Transfer the block production reward for this block proposer:
-            increase_value_for_key(&mut result, block.proposer().clone(), production_reward)?;
+            increase_value_for_key(block.proposer.clone(), production_reward)?;
 
             // Now, let's compute the reward attached to each signed block reported by the block
             // we examine:
             for (signature_rewards, signed_block_height) in block
-                .rewarded_signatures()
+                .rewarded_signatures
                 .iter()
-                .zip((0..block.height()).rev())
+                .zip((0..block.height).rev())
             {
-                let signed_block_era = cited_blocks
-                    .iter()
-                    .flatten()
-                    .find_map(|block| {
-                        (block.height() == signed_block_height).then_some(block.era_id())
-                    })
-                    .ok_or_else(|| RewardsError::HeightNotInEraRange(block.height()))?;
-                let validators_providing_signature =
-                    signature_rewards.to_validator_set(eras_info.validator_keys(signed_block_era)?);
+                let signed_block_era = rewards_info.era_for_block_height(signed_block_height)?;
+                let validators_providing_signature = signature_rewards
+                    .to_validator_set(rewards_info.validator_keys(signed_block_era)?);
 
                 for signing_validator in validators_providing_signature {
                     // Reward for contributing to the finality signature, ie signing this block:
-                    let contribution_reward = eras_info
+                    let contribution_reward = rewards_info
                         .weight_ratio(signed_block_era, &signing_validator)?
                         .checked_mul(&contribution_proportion)
                         .ok_or(RewardsError::ArithmeticOverflow)?
-                        .checked_mul(&eras_info.reward(signed_block_era)?)
+                        .checked_mul(&rewards_info.reward(signed_block_era)?)
                         .ok_or(RewardsError::ArithmeticOverflow)?;
                     // Reward for gathering this signature. It is both weighted by the block
                     // producing/signature collecting validator, and the signing validator:
-                    let collection_reward = eras_info
+                    let collection_reward = rewards_info
                         .weight_ratio(signed_block_era, &signing_validator)?
                         .checked_mul(&collection_proportion)
                         .ok_or(RewardsError::ArithmeticOverflow)?
-                        .checked_mul(&eras_info.reward(signed_block_era)?)
+                        .checked_mul(&rewards_info.reward(signed_block_era)?)
```

Comment:
Should collection be multiplied by the signed block era reward, or by the current block era reward? It was collected in _this_ era, after all, by a validator from _this_ era.

Reply:
I think it makes it somewhat easier to describe and keep the calculations straight if the reward is based on the era of the signed block, and it's easy to see what the expected values come out to be

Reply:
I checked with Alex, it's the right weighting.